docs/dep/dep-006-new-go-sdk.md

```diff
@@ -0,0 +1,474 @@
+
+<a id="callouts"></a>
+
+## Callouts
+
+* This document covers ONLY how to create new components via SDK
+* The advertised proposal will be worked on incrementally
+* Environment aware actions/components will be discussed after this document to keep the scope understandable and concise
+* Some more technical bits about retries, metrics and so on have been left out for the moment.
+  We'll discuss them as a follow-up in RFCs.
+* The implementation will likely differ a bit from the proposal as we'll tweak the final the SDK
+  while migrating components. We'll track the various changes to the SDK with RFCs.
+
+<a id="proposal"></a>
+
+## Proposal
+
+<a id="proposal-structure"></a>
+
+### Structure and contributions guidelines
+
+The code for the new SDK will live in this repository in the `sdk/component` directory.
+
+The SDK will have its own go module `github.com/ocurity/dracon/sdk/component` with its own set of dependencies.
+
+> Why?
+>
+> * having a separate directory helps keeping responsibilities segregated
+> * having a separate module helps keeping dependencies separate, we want very few dependencies on the core SDK to boost adoption
+> * it's trivial to move the SDK in its own repository, if needed
+> * independent versioning from dracon-oss
+
+The structure of the files will follow these guidelines:
+
+```
+dracon/
+├── ...
+├── sdk/
+│   └── component/
+│       ├── go.mod
+│       ├── component.go
+│       ├── reader.go
+│       ├── writer.go
+│       ├── transformer.go
+│       ├── ...
+│       └── examples/
+└── ...
+
+** The file names are an example
+```
+
+All the files belonging to the SDK will leave on the root level of the package to simplify import paths.
+
+Unexported code MUST be unexported or live inside the `internal/` directory.
+
+<a id="proposal-component-api"></a>
+
+### Component API
+
+At its core, regardless of the type of the component, each one of them has to perform 3 actions:
+
+* `Read`: read raw information from some source - filesystem, database, api, ...
+* `Transform`: perform operations to enrich the raw data based on the context - adding annotations, call an api, ...
+* `Write`: write the enriched results - filesystem, database, api ...
+
+Different components implement these actions in one way or another. Some of them can be NOOP.
+
+The components share information between different actions. This information follows a predictable format of a vulnerability report.
+
+This means that we can provide a minimum set of API to guide the creation of new components. The Component API will be represented by a `Runner` interface which can be satisfied to implement a new component:
+
+```go
+// Runner advertises the actions that need to be implemented by a Smithy component.
+type Runner interface {
+    // Read should be implemented to read Smithy's raw input from an underlying data source.
+    Read(ctx context.Context) (*v1.VulnerabilityFinding, error)
```

Comment:
The runner looks great for vulnerability scanners, but some scanners return other data, like SBOMs. Would they need to implement a different runner? Or should it be something like "DataFinding"?

Reply:
My understanding is that right now we want to get started with [VulnerabilityFinding](https://schema.ocsf.io/1.3.0/classes/vulnerability_finding) from [oscf](https://schema.ocsf.io/) and this can be used to represent all the inflight data that we need between components (cc @northdpole )

In any case, if we wanted to cover other categories, we could easily do so by leveraging an enum

I think for now this is good enough but let me know what you think!

Reply:
I agree with @andream16 , for now, VulnFinding is good enough, for communicating sboms, since ocsf doesn't give you any way to attach a cyclonedx doc, I think we can use the ["observables"](https://schema.ocsf.io/1.3.0/objects/observable?extensions=) array to communicate a serialized Cyclonedx document